tests/unit/test-class-wc-payment-gateway-wcpay.php

```diff
@@ -1992,6 +1994,50 @@ public function test_process_payment_caches_mimimum_amount_and_displays_error_up
 		}
 	}
 
+	public function test_process_payment_rejects_if_missing_fraud_prevention_token() {
+		$order = WC_Helper_Order::create_order();
+
+		$fraud_prevention_service_mock = $this->getMockBuilder( Fraud_Prevention_Service::class )
+			->disableOriginalConstructor()
+			->getMock();
+		Fraud_Prevention_Service::set_instance( $fraud_prevention_service_mock );
+
+		$fraud_prevention_service_mock
+			->expects( $this->once() )
+			->method( 'is_enabled' )
+			->willReturn( true );
+
+		$this->expectException( Exception::class );
+		$this->expectExceptionMessage( 'Your payment was not processed.' );
+		$this->wcpay_gateway->process_payment( $order );
+	}
+
+	public function test_process_payment_rejects_if_invalid_fraud_prevention_token() {
+		$order = WC_Helper_Order::create_order();
+
+		$fraud_prevention_service_mock = $this->getMockBuilder( Fraud_Prevention_Service::class )
+			->disableOriginalConstructor()
+			->getMock();
+		Fraud_Prevention_Service::set_instance( $fraud_prevention_service_mock );
+
+		$fraud_prevention_service_mock
+			->expects( $this->once() )
+			->method( 'is_enabled' )
+			->willReturn( true );
+
+		$fraud_prevention_service_mock
+			->expects( $this->once() )
+			->method( 'verify_token' )
+			->with( 'incorrect-token' )
+			->willReturn( false );
+
+		$_POST['wcpay-fraud-prevention-token'] = 'incorrect-token';
+
+		$this->expectException( Exception::class );
+		$this->expectExceptionMessage( 'Your payment was not processed.' );
+		$this->wcpay_gateway->process_payment( $order );
+	}
+
```

Comment:
I feel that this is missing the "success case" but then there are 2 options:
- Setting everything up for success case. It's a very complicated setup and seems to be the reason there isn't a test covering that so far.
- create a partial mock for gateway class and assert that `process_payment_for_order` is called (passed the fraud prevention checks). I gave it a try and say that the setup is complicated here as well since we need to mock things until that method like ` $this->failed_transaction_rate_limiter->is_limited()` 
- Make ` $this->failed_transaction_rate_limiter->is_limited()` fail and checking that method fail because of rate limiter. This seems doable but then the test relies on the order of execution. Don't know if that's a major issue.

Reply:
> Ideally, I'd like to assert exception code as well but it seems impossible.

Correct, that's due to the custom exception layer.

> Is there a reason we shouldn't add the error code when re-throwing the exception?

Looking into codebase and git history, no clues on my end.

Reply:
I like the second option (`create a partial mock for gateway class ...`), can you give it a try.

Reply:
Implemented in [883ee4b06e0deeddb8adc7dc1e7358f33b7aa049](https://github.com/Automattic/woocommerce-payments/pull/3979/commits/883ee4b06e0deeddb8adc7dc1e7358f33b7aa049)

Reply:
Look good!